### PR TITLE
Migration to typescript, removal of partition processors

### PR DESCRIPTION
**Pull request by jsnoble**
- converts assets to use typescript
- removes partition processors in favor of routers in standard assets

### CODE REVIEW

File: asset/src/__lib/chunked-file-reader/formatters.ts [hunk 1]
```diff
@@ -92,15 +92,18 @@ function csv(incomingData, logger, opConfig, metadata, slice) {
...
-    });
+    }
+
+    return Promise.all(data.map(call));
 }
 
 // tsv is just a specific case of csv
-function tsv(incomingData, logger, opConfig, metadata, slice) {
-    return csv(incomingData, logger, opConfig, metadata, slice);
+export function tsv(incomingData: any, logger: Logger, opConfig: any, metadata: any, slice: any) {
```
Comment (peterdemartini) [inline asset/src/__lib/chunked-file-reader/formatters.ts]: What is the point of using typescript we use `any` everywhere?
```suggestion
export function tsv(incomingData: string, logger: Logger, opConfig: FileConfig, metadata: DataEntityMetadata, slice: FIXME) {
```

File: asset/src/__lib/chunked-file-reader/formatters.ts [hunk 2]
```diff
@@ -50,7 +48,7 @@ function raw(incomingData, logger, opConfig, metadata, slice) {
     });
 }
 
-function csv(incomingData, logger, opConfig, metadata, slice) {
+export function csv(incomingData: any, logger: Logger, opConfig: any, metadata: any, slice: any) {
```
Comment (peterdemartini) [inline asset/src/__lib/chunked-file-reader/formatters.ts]:
```suggestion
export function csv(incomingData: string, logger: Logger, opConfig: FileConfig, metadata: DataEntityMetadata, slice: FIXME) {
```

File: asset/src/__lib/chunked-file-reader/formatters.ts [hunk 3]
```diff
@@ -31,9 +28,10 @@ function _toRecords(rawData, delimiter, slice) {
 }
 
 // No parsing, leaving to reader or a downstream op.
-function raw(incomingData, logger, opConfig, metadata, slice) {
+export function raw(incomingData: any, logger: Logger, opConfig: any, metadata: any, slice: any) {
```
Comment (peterdemartini) [inline asset/src/__lib/chunked-file-reader/formatters.ts]:
```suggestion
export function raw(incomingData: string, logger: Logger, opConfig: FileConfig, metadata: DataEntityMetadata, slice: FIXME) {
```

File: asset/src/__lib/chunked-file-reader/formatters.ts [hunk 4]
```diff
@@ -1,13 +1,10 @@
...
-const csvToJson = require('csvtojson');
-const { DataEntity } = require('@terascope/utils');
+import csvToJson from 'csvtojson';
+import { DataEntity, Logger } from '@terascope/utils';
 
 // This function takes the raw data and breaks it into records, getting rid
 // of anything preceding the first complete record if the data does not
 // start with a complete record.
-function _toRecords(rawData, delimiter, slice) {
+function _toRecords(rawData: string, delimiter: string, slice: any) {
```
Comment (peterdemartini) [inline asset/src/__lib/chunked-file-reader/formatters.ts]:
```suggestion
function _toRecords(rawData: string, delimiter: string, slice: FIXME) {
```

File: asset/src/__lib/chunked-file-reader/formatters.ts [hunk 5]
```diff
@@ -136,9 +139,11 @@ function json(incomingData, logger, opConfig, metadata) {
     }
 }
 
-function ldjson(incomingData, logger, opConfig, metadata, slice) {
+export function ldjson(
+    incomingData: any, logger: Logger, opConfig: any, metadata: any, slice: any
+) {
     const data = _toRecords(incomingData, opConfig.line_delimiter, slice);
-    return data.map((record) => {
+    return data.map((record: any) => {
```
Comment (peterdemartini) [inline asset/src/__lib/chunked-file-reader/formatters.ts]: Adding this `(record: any)`  isn't needed if proper types are returned from `_toRecords`

File: asset/src/hdfs_reader/slicer.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    /**
+     * Currently only enable autorecover jobs
+     *
+     * @todo we should probably support full recovery
+    */
+    isRecoverable() {
+        return Boolean(this.executionConfig.autorecover);
+    }
+
+    searchFiles(metadata: any, filePath: string) {
```
Comment (peterdemartini) [inline asset/src/hdfs_reader/slicer.ts]: Should we define a type for `metadata`?

File: asset/src/__lib/chunked-file-reader/index.ts [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+        currentOffset += length;
+    }
+    // Don't read too far - next slice will get it.
+    return margin.split(delimiter)[0];
+}
+
+// This function will grab the chunk of data specified by the slice plus an
+// extra margin if the slice does not end with the delimiter.
+export async function getChunk(
+    readerClient: any, slice: any, opConfig: any, logger: any, metadata: any
```
Comment (peterdemartini) [inline asset/src/__lib/chunked-file-reader/index.ts]: We should be more specific on the types here

File: asset/src/__lib/chunked-file-reader/index.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+    const lastChunk = total % size;
+
+    if (lastChunk > 0) {
+        chunks.push({ offset: (fullChunks * size) - delta, length: lastChunk + delta });
+    }
+
+    return chunks;
+}
+
+async function getMargin(readerClient: any, delimiter: string, offset: number, length: number) {
```
Comment (peterdemartini) [inline asset/src/__lib/chunked-file-reader/index.ts]: We need a type for `readerClient`

**Comment by kstaken** [comment]
Unfortunately the routing behavior in this doesn't work. It just results in `TSError: The specified bucket does not exist` exceptions.

Previously the partitioners would generate the entire path that would then be written by the sender. 

So previously the senders were expecting `file:partition=/base/path/bucket` but now will get `standard:route=bucket`.

What we had discussed doing was taking the `path` configuration from the sender and simply combining that with `standard:route`. So `path=/some/base/path` when there is an associated `standard:route=bucket1` would result in a path of `/some/base/path/bucket1` and then all files would be collected under that path. 

If `standard:route=1` like might be the output from `hash_router` then the path would be `/some/base/path/1`.

NOTE: this path is the directory, NOT the file name. Workers will each be creating files under that path using the existing random filenames.

**Comment by jsnoble** [comment]
Example job Routing from ES to File
```
{
    "name": "s3_routing",
    "lifecycle": "once",
    "analytics": true,
    "slicers": 1,
    "workers": 1,
    "assets": [
        "file",
        "standard",
        "elasticsearch"
    ],
    "operations": [
        {
            "_op": "elasticsearch_reader",
            "size": 500,
            "index": "id_test_index",
            "type": "events",
            "date_field_name": "created",
            "time_resolution": "ms"
        },
        {
            "_op": "key_router",
            "from": "beginning",
            "use": 1
        },
        {
            "_op": "file_exporter",
            "path": "/Users/jarednoble/Desktop/fs/test",
            "format": "ldjson",
            "file_per_slice": true
        }
    ]
}
```

Example job routing ES to S3
```
{
    "name": "s3_routing",
    "lifecycle": "once",
    "analytics": true,
    "slicers": 1,
    "workers": 1,
    "assets": [
        "file",
        "standard",
        "elasticsearch"
    ],
    "operations": [
        {
            "_op": "elasticsearch_reader",
            "size": 500,
            "index": "id_test_index",
            "type": "events",
            "date_field_name": "created",
            "time_resolution": "ms"
        },
        {
            "_op": "key_router",
            "from": "beginning",
            "use": 1
        },
        {
            "_op": "s3_exporter",
            "path": "routing",
            "file_per_slice": true
        }
    ]
}
```

**Comment by kstaken** [comment]
Senders with routing look pretty good now.

Unfortunately it looks like reading doesn't work now if there are subdirectories. Reading should recurse and read all data in the hierarchy. 

```(node:39994) UnhandledPromiseRejectionWarning: TSError: The specified key does not exist.
    at Object.pRetry (/Users/kstaken/projects/customer/spaces/activities/node_modules/@terascope/utils/dist/src/promises.js:55:21)
    at Request.extractError (/Users/kstaken/projects/customer/spaces/activities/node_modules/aws-sdk/lib/services/s3.js:816:35)
    at Request.callListeners (/Users/kstaken/projects/customer/spaces/activities/node_modules/aws-sdk/lib/sequential_executor.js:106:20)
    at Request.emit (/Users/kstaken/projects/customer/spaces/activities/node_modules/aws-sdk/lib/sequential_executor.js:78:10)
    at Request.emit (/Users/kstaken/projects/customer/spaces/activities/node_modules/aws-sdk/lib/request.js:683:14)
    at Request.transition (/Users/kstaken/projects/customer/spaces/activities/node_modules/aws-sdk/lib/request.js:22:10)
```

**Pull request merged**